### PR TITLE
[temp.over.link] Reword to clarify that declarations correspond

### PR DESCRIPTION
**Pull request by tkoeppe**
Fixes #https://github.com/cplusplus/draft/issues/5997.

### CODE REVIEW

**Comment by tkoeppe** [comment]
@burblebee, @opensdh, @jensmaurer: For your consideration, please!

**Comment by W-E-Brown** [comment]
I just happened to glance at this draft rewording:  I'm sorry, but I can't understand the proposed sentence.  Please revise, preferably into two or more sentences; this one as drafted just has too many convolutions for me to decipher.  Failing that, please restore the original.

**Comment by tkoeppe** [comment]
@W-E-Brown: Could you please tell me how you understand the original wording? I also find this a bit weird, but I don't think this change is at fault here -- the problem for me already resided in the original third bullet, "would declare the same entity considering them to correspond". I thought I'm just preserving that weirdness. If you tell me what the original meant, I can try and look for a clearer rewording.

**Comment by W-E-Brown** [comment]
IMO, "preserving that weirdness" seems to be a questionable objective of rewording an admittedly questionable specification.  Alas, I see no improvement in clarity in the proposed text.

Having any "weirdness" in a core specification seems strong motivation for consulting CWG or even opening a core issue:  I suggest that CWG be asked to clarify the intention and its wording here, since "considering them to correspond" is (AFAIK) not a defined term of art.  Pending resolution of such inquiry, I respectfully recommend no wording change, as I find the proposed bullet-free specification much harder to parse than the existing specification with its bullet list.

**Comment by tkoeppe** [comment]
The wording came in fresh  out of Core with CWG2603, and I wouldn't normally second-guess Core's deliberations and care. @jensmaurer, @opensdh Could you perhaps shed some light on how "considering them to correspond" is to be interpreted?

**Comment by jensmaurer** [comment]
The idea of that bullet is to do a "specification function call" to "same entity" with the override "the declarations do correspond" (although, in fact, they don't).

"Same entity" says in 6.6 p8:

> Two declarations of entities declare the same entity if, considering declarations of unnamed types to introduce their names for linkage purposes, if any (9.2.4, 9.7.1), they correspond (6.4.1), have the same target scope that is not a function or template parameter scope, and either ...

We want to have this list of conditions checked, except that "they correspond" should be assumed to be true for that specification function all.

"considering them to correspond" seems to be on-point for that.

**Comment by tkoeppe** [comment]
OK, I get that, but I still think "if two function templates [...] would declare the same entity considering them to correspond" is a pretty awkward way to say that. I think the fact that the "them" refers to something quite distant, and also to the fact that you're expected to spot that "correspond" is a detail of "same entity" make this quite a strain.

In other words, this wording makes sense if you already have your explanation in mind, but the explanation itself seems hard to discover just based off this wording.

Would it hurt to just spell this reasoning out?

"""
* would declare the same entity,<ins> where in the determination in \ref{basic.link} of whether two declarations do declare the same entity, the declarations are assumed to correspond,</ins> and
"""

**Comment by tkoeppe** [comment]
Maybe we can decorate the whole sentence a bit more:

""
Furthermore, if two declarations $A$ and $B$ of function templates that do not correspond
* declare the same name,
* have corresponding signatures,
* would declare the same entity, where in the determination in \ref{basic.link} of whether two declarations do declare the same entity, $A$ and $B$ are assumed to correspond, and
* accept and are satisfied by the same set of template argument lists,

the program is ill-formed [...]
"""

**Comment by tkoeppe** [comment]
We could also move the punchline (that the declarations do not correspond) to the end:

"""
Furthermore, if two declarations $A$ and $B$ of function templates [List] but do not correspond, the program is ill-formed [...]
"""

**Comment by tkoeppe** [comment]
image

@jensmaurer, @opensdh What do you think of this suggestion?

**Comment by jensmaurer** [comment]
I like moving the punchline.  I'm not too happy with that third bullet; it feels too wordy.

maybe

"would declare the same entity, when considering A and B to correspond in the determination per 6.6,"

**Comment by tkoeppe** [comment]
Done:

image

**Comment by jensmaurer** [comment]
@opensdh, what do you think?

**Comment by tkoeppe** [comment]
Even more radical, and I'm not actually proposing that:

===
The program is ill-formed, no diagnostic required, if

   * the validity or meaning of the program [...] and they are functionally equivalent but not equivalent, or
   * two declarations $A$ and $B$ do not correspond, but they
       * declare the same name,
       * have corresponding signatures,
       * would declare the same entity [...], and
       * accept and are satisfied by the same set of template argument lists.

===

**Comment by opensdh** [comment]
I like 58b98d1426cdbe6dccb49424d8b30393fe03ee87 (though perhaps with some slight rewordings like "declare the same name" &rarr; "introduce the same name" and "to correspond in the determination per 6.6," &rarr; "in that determination (6.6),").  The whole point here is to recompute the same-entity status with a deliberately weaker filter just as with "functionally equivalent but not equivalent".

**Review by jensmaurer — APPROVED** [review]
Works for me.

**Comment by tkoeppe** [comment]
Thanks, everyone!
